### PR TITLE
Variables in allowed position

### PR DESCRIPTION
**Pull request by rawls238**
trying to figure out why this test has been failing...

@jhgg @dittos @woodb let me know if you have any ideas, I've been going down a few rabbit holes in the last couple hours :)

### CODE REVIEW

File: graphql/core/validation/rules.py [hunk 1]
```diff
@@ -591,7 +591,54 @@ def bad_value_for_default_arg_message(var_name, type, value):
...
+    def enter_OperationDefinition(self, *args):
+        self.var_def_map = {}
+        self.visited_fragment_names = set()
+
+    def enter_VariableDefinition(self, var_def_ast, *args):
+        self.var_def_map[var_def_ast.variable.name.value] = var_def_ast
+
+    def enter_Variable(self, variable_ast, *args):
+        var_name = variable_ast.name.value
+        var_def = self.var_def_map[var_name]
```
Comment (rawls238) [inline graphql/core/validation/rules.py]: the test fails since booleanArg isn't a key in this map.

**Comment by jhgg** [comment]
I'll take a look later.

**Comment by woodb** [comment]
So I spent a lot of time trying to figure this one out, and ending up having to read so much code lol... And no solution. I want to say it's some weird language difference between JavaScript and Python that is causing this error... Someone with more intimate knowledge of all this code (ahem, @dittos :)) might be able to zap it. Best of luck to anyone, I give up!

File: graphql/core/validation/rules.py (hunk 1 above)
Comment (dittos) [inline graphql/core/validation/rules.py]: Set `var_def` to `None` if value is not found. That's the way graphql-js works :grinning:

**Comment by jhgg** [comment]
I've made the fixes in https://github.com/rawls238/graphql-py/pull/1

Was a simple issue really (ignore the formatting fixes)

In the JS implementation:

```
var varDef = varDefMap[varName];
var varType = varDef && typeFromAST(context.getSchema(), varDef.type);
```

Remember that accessing a key that doesn't exist in a JS object returns `undefined`, but in Python it will raise a `KeyError`, so instead, we will use `.get(var_name)`.

File: graphql/core/validation/rules.py (hunk 1 above)
Comment (jhgg) [inline graphql/core/validation/rules.py]: Just read this comment, but essentially same as what I said in https://github.com/dittos/graphqllib/pull/38#issuecomment-141724009 !

**Comment by jhgg** [comment]
:+1: after this one's merged, 2 more to go!

File: graphql/core/validation/rules.py [hunk 2]
```diff
@@ -591,7 +591,64 @@ def bad_value_for_default_arg_message(var_name, type, value):
...
+        self.visited_fragment_names.add(node.name.value)
+
+    @staticmethod
+    def effective_type(var_type, var_def):
+        if not var_def.default_value or isinstance(var_def, GraphQLNonNull):
+            return var_type
+
+        return GraphQLNonNull(var_type)
+
+    @staticmethod
```
Comment (dittos) [inline graphql/core/validation/rules.py]: How about changing this to `@classmethod` and shorten `VariablesInAllowedPosition` below to `cls`?

Comment (jhgg) [inline graphql/core/validation/rules.py]: :+1:

**Comment by rawls238** [comment]
@jhgg @dittos same as the other branch, going to try to come back to this on Tues / Wednesday but there is effectively one failing test still.
```
complicatedArgs {
         stringListArgField(stringListArg: [$stringVar])
}
```

appears to be trying ```[$stringVar]``` as a String! instead of a [String!] for some reason which is causing the tests to fail

**Comment by jhgg** [comment]
I'll take a look!

**Comment by dittos** [comment]
There was a bug in the visitor. https://github.com/dittos/graphqllib/commit/045390b9b0f51ba29fb4043c427ec1eb47ec5183 It will work properly if you merge or rebase on that change!

**Comment by rawls238** [comment]
thanks @dittos, seems to have fixed the issue. This should be ready to go

**Comment by dittos** [comment]
2 rules to go!

**Pull request merged**